In [1]:
from itertools import product
from fractions import Fraction

def generation_tuples(n):
    """
    Enumerate all (label, (a,b,c)) tuples for generation n.

    Parameters
    ----------
    n : int
        Generation number (length of label).

    Yields
    ------
    label : str
        The binary string label of length n.
    triple : tuple(int, int, int)
        The associated (a,b,c) triple.
    """
    seqs = product('01', repeat=n)

    for bits in seqs:
        label = ''.join(bits)
        ones = [i+1 for i, b in enumerate(bits) if b == '1']
        k = len(ones)
        # compute c = sum_{j=1}^{k} 3^{k-j} * 2^{i_j - 1}
        c = sum((3 ** (k - j - 1)) * (2 ** (i - 1)) for j, i in enumerate(ones))
        yield label, (n, k, c)
#


In [2]:
def CollatzChain(collatzNumber):
    chain = [collatzNumber]
    while collatzNumber != 1:
        if collatzNumber & 1 == 0:
            collatzNumber = collatzNumber // 2
        else:
            collatzNumber = (3 * collatzNumber + 1) // 2
        chain.append(collatzNumber)
    return chain
#

def ChainPath(collatzNumber):
    path = []
    while collatzNumber != 1:
        if (collatzNumber & 1) == 0:
            collatzNumber = collatzNumber // 2
            path.append("1")
        else:
            collatzNumber = (3 * collatzNumber + 1) // 2
            path.append("0")
    return "".join(path)
#
def fractionFromNodeTup(tup):
    p2, p3, c = tup
    fract = Fraction(2**p2 - c, 3**p3)
    return (fract.numerator, fract.denominator)
#

def TupChainFromPath(chain_path):
    tup_chain = [(0, 0, 0)]
    for chain_item in chain_path:
        p2, p3, c = tup_chain[-1]
        if chain_item == "1":
            tup_chain. append((p2 + 1, p3, c))
        else:
            tup_chain. append((p2 + 1, p3 + 1, c*3 + 2**p2))
            
        fract = fractionFromNodeTup(tup_chain[-1])
    return tup_chain
#
def FractionFromPath(chain_path):
    tup = (0, 0, 0)
    for chain_item in chain_path:
        p2, p3, c = tup
        if chain_item == "1":
            tup = (p2 + 1, p3, c)
        else:
            tup = (p2 + 1, p3 + 1, c*3 + 2**p2)
        
    fract = fractionFromNodeTup(tup)
    return fract
#    
def downUpTup(tup):
    p2, p3, c = tup
    p2_01, p3_01, c_01 = (p2 + 2, p3 + 1, c*3 + 2**p2)
    return (p2_01, p3_01, c_01)
#
def TupChain(collatzNumber):
    chain_path = ChainPath(collatzNumber)
    return TupChainFromPath(chain_path)
    return tup_chain
#
def generationTups(n):
    G = [[(0, 0, 0)]]
    for i in range(1, n+1, 1):
        G.append([])
        tups = G[i-1]
        for tup in tups:
            p2, p3, c = tup
            G[i].extend([(p2 + 1, p3, c), (p2 + 1, p3 + 1, c*3 + 2**p2)])
    return G
#
def pathToTup(path):
    a = len(path)
    zeros = [i for i, b in enumerate(path) if b == '0']
    b = len(zeros)
    # compute c = sum_{j=1}^{k} 3^{k-j} * 2^{i_j - 1}
    c = sum((3 ** (b - j - 1)) * (2 ** (i)) for j, i in enumerate(zeros))
    fract = Fraction(2**a - c, 3**b)
    return path, (a, b, c), (fract.numerator, fract.denominator)
#
def reverseColllatz(collatz_num, generations):
    chains = [[collatz_num]]
    for i in range(generations):
        for chain in chains.copy():
            if ((chain[-1] -1) % 3) == 0:
                reverse_down_val = (chain[-1] - 1)//3
                if (reverse_down_val & 1) == 1:
                    # down only applies to odds
                    chain2 = chain.copy()
                    chain2.append(reverse_down_val)
                    chains.append(chain2)
            chain = chain.append(2*chain[-1])
    #
    return chains
#


def collatzNext(x):
    if x %2 == 0:
        return x//2
    return 3*x + 1
#
def fourPlusOneEdges(maxN):
    for n in range(maxN):
        start = 4*n + 1
        y = collatzNext(start)
        while y % 4 != 1:
            y = collatzNext(y)
        print((n, (y-1)//4))
#
def ToNPlusOneEdges(g, maxN):
    L = []
    multiplier = 2**g
    for n in range(maxN):
        start = multiplier*n + 1
        y = collatzNext(start)
        while y % multiplier != 1:
            y = collatzNext(y)
        L.append((n, (y-1)//multiplier))
    return L
#


def two_factors(n):
    """
    Returns the exponent of the largest power of 2 that divides n.
    For example, if n = 48 (binary 110000), the result is 4 because 2^4 = 16 divides 48.
    """
    if n <= 1:
        return 0 

    count = 0
    while (n & 1) == 0:
        n >>= 1
        count += 1
    return count
#

In [3]:
TupChain(27)[-1]

(70, 41, 195820718533800070543)

In [4]:
for tup in generation_tuples(20):
    p2, p3, c = tup[1]
    fract = Fraction(2**p2 - c, 3**p3)
    if (fract.denominator == 1):
        print((tup[0], fract.numerator))

('00000000000000000000', 1048576)
('00000000000000000010', 262144)
('00000000000000001000', 327680)
('00000000000000001010', 65536)
('00000000000000011000', 98304)
('00000000000000100000', 344064)
('00000000000000100010', 81920)
('00000000000000101010', 16384)
('00000000000001001000', 106496)
('00000000000001100010', 24576)
('00000000000010000000', 348160)
('00000000000010000010', 86016)
('00000000000010001010', 20480)
('00000000000010101010', 4096)
('00000000000100001000', 108544)
('00000000000100100010', 26624)
('00000000000101001000', 34816)
('00000000000110001010', 6144)
('00000000001000000000', 349184)
('00000000001000000010', 87040)
('00000000001000001010', 21504)
('00000000001000101010', 5120)
('00000000001010000000', 115712)
('00000000001010101010', 1024)
('00000000001100001000', 35840)
('00000000001101001000', 11264)
('00000000010000001000', 109056)
('00000000010000100010', 27136)
('00000000010001001000', 35328)
('00000000010010001010', 6656)
('00000000010100100010', 8704)
('0

In [5]:
def generation_tuples2(n):
    """
    Enumerate all (label, (a,b,c)) tuples for generation n.  Giving labels consistent with other notebooks

    Parameters
    ----------
    n : int
        Generation number (length of label).

    Yields
    ------
    label : str
        The binary string label of length n.
    triple : tuple(int, int, int)
        The associated (a,b,c) triple.
    """
    seqs = product('10', repeat=n)

    for bits in seqs:
        label = ''.join(bits)
        zeros = [i for i, b in enumerate(bits) if b == '0']
        k = len(zeros)
        # compute c = sum_{j=0}^{k} 3^{k-j} * 2^{i_j - 1}
        c = sum((3 ** (k - j - 1)) * (2 ** (i)) for j, i in enumerate(zeros))
        yield label, (n, k, c)
#

In [6]:
for tup in generation_tuples2(20):
    p2, p3, c = tup[1]
    fract = Fraction(2**p2 - c, 3**p3)
    if (fract.denominator == 1):
        print((tup[0], fract.numerator))

('11111111111111111111', 1048576)
('11111111111111111101', 262144)
('11111111111111110111', 327680)
('11111111111111110101', 65536)
('11111111111111100111', 98304)
('11111111111111011111', 344064)
('11111111111111011101', 81920)
('11111111111111010101', 16384)
('11111111111110110111', 106496)
('11111111111110011101', 24576)
('11111111111101111111', 348160)
('11111111111101111101', 86016)
('11111111111101110101', 20480)
('11111111111101010101', 4096)
('11111111111011110111', 108544)
('11111111111011011101', 26624)
('11111111111010110111', 34816)
('11111111111001110101', 6144)
('11111111110111111111', 349184)
('11111111110111111101', 87040)
('11111111110111110101', 21504)
('11111111110111010101', 5120)
('11111111110101111111', 115712)
('11111111110101010101', 1024)
('11111111110011110111', 35840)
('11111111110010110111', 11264)
('11111111101111110111', 109056)
('11111111101111011101', 27136)
('11111111101110110111', 35328)
('11111111101101110101', 6656)
('11111111101011011101', 8704)
('1

In [7]:
list(enumerate('000011110111'))
    

[(0, '0'),
 (1, '0'),
 (2, '0'),
 (3, '0'),
 (4, '1'),
 (5, '1'),
 (6, '1'),
 (7, '1'),
 (8, '0'),
 (9, '1'),
 (10, '1'),
 (11, '1')]

In [8]:
#def pathToTup(path):
#    g = len(path)
#    zeros = [i+1 for i, b in enumerate(path) if b == '0']
#    k = len(zeros)
#    # compute c = sum_{j=1}^{k} 3^{k-j} * 2^{i_j - 1}
#    c = sum((3 ** (k - j - 1)) * (2 ** (i - 1)) for j, i in enumerate(zeros))
#    fract = Fraction(2**g - c, 3**k)
#    return path, (g, k, c), (fract.numerator, fract.denominator)
    

In [9]:
pathToTup('000011110111')

('000011110111', (12, 5, 451), (15, 1))

In [10]:
def pathToTup(path):
    a = len(path)
    zeros = [i for i, b in enumerate(path) if b == '0']
    b = len(zeros)
    # compute c = sum_{j=1}^{k} 3^{k-j} * 2^{i_j - 1}
    c = sum((3 ** (b - j - 1)) * (2 ** (i)) for j, i in enumerate(zeros))
    fract = Fraction(2**a - c, 3**b)
    return path, (a, b, c), (fract.numerator, fract.denominator)

In [11]:
pathToTup('000011110111')

('000011110111', (12, 5, 451), (15, 1))

In [33]:
for tup in generation_tuples2(3):
    p2, p3, c = tup[1]
    fract = Fraction(2**p2 - c, 3**p3)
    print((tup[0], (p2, p3, c), (fract.numerator, fract.denominator)))

('111', (3, 0, 0), (8, 1))
('110', (3, 1, 4), (4, 3))
('101', (3, 1, 2), (2, 1))
('100', (3, 2, 10), (-2, 9))
('011', (3, 1, 1), (7, 3))
('010', (3, 2, 7), (1, 9))
('001', (3, 2, 5), (1, 3))
('000', (3, 3, 19), (-11, 27))


In [12]:
for tup in generation_tuples2(7):
    p2, p3, c = tup[1]
    fract = Fraction(2**p2 - c, 3**p3)
    print((tup[0], (p2, p3, c), (fract.numerator, fract.denominator)))

('1111111', (7, 0, 0), (128, 1))
('1111110', (7, 1, 64), (64, 3))
('1111101', (7, 1, 32), (32, 1))
('1111100', (7, 2, 160), (-32, 9))
('1111011', (7, 1, 16), (112, 3))
('1111010', (7, 2, 112), (16, 9))
('1111001', (7, 2, 80), (16, 3))
('1111000', (7, 3, 304), (-176, 27))
('1110111', (7, 1, 8), (40, 1))
('1110110', (7, 2, 88), (40, 9))
('1110101', (7, 2, 56), (8, 1))
('1110100', (7, 3, 232), (-104, 27))
('1110011', (7, 2, 40), (88, 9))
('1110010', (7, 3, 184), (-56, 27))
('1110001', (7, 3, 152), (-8, 9))
('1110000', (7, 4, 520), (-392, 81))
('1101111', (7, 1, 4), (124, 3))
('1101110', (7, 2, 76), (52, 9))
('1101101', (7, 2, 44), (28, 3))
('1101100', (7, 3, 196), (-68, 27))
('1101011', (7, 2, 28), (100, 9))
('1101010', (7, 3, 148), (-20, 27))
('1101001', (7, 3, 116), (4, 9))
('1101000', (7, 4, 412), (-284, 81))
('1100111', (7, 2, 20), (12, 1))
('1100110', (7, 3, 124), (4, 27))
('1100101', (7, 3, 92), (4, 3))
('1100100', (7, 4, 340), (-212, 81))
('1100011', (7, 3, 76), (52, 27))
('1100010

In [13]:
# (7, 7, 2059)
(2**7, 3**7, 2059)

(128, 2187, 2059)

In [14]:
2187 - 2059

128

In [15]:
def NecklacePeers(path):
    paths = [path]
    a = len(path)
    for i in range(a):
        head = paths[-1][0]
        tail = paths[-1][1:]
        rot_path = tail + head
        if rot_path == path:
            break
        paths.append(rot_path)
    return sorted(paths)
#

def CollatzNumNecklaces(collatz_num, num_gen):
    path = ChainPath(collatz_num)
    a = len(path)
    generation_paths = {}
    for i in range(num_gen):
        paths = NecklacePeers(path)
        generation_paths[a] = paths
        a +=2
        path = path + "01"
    #
    return generation_paths
#

def strip01suffixes(path):
    while (len(path) > 2) and (path[-2:] == "01"):
        path = path[0:-2]
    return path
#
def isNew(path):
    if len(path) == len(strip01suffixes(path)):
        return True
    return False
#

def CollatzNumNecklacesMeta(collatz_num, num_gen, filter_new = False, filter_int = False):
    meta = {}
    generation_paths = CollatzNumNecklaces(collatz_num, num_gen)
    for key in generation_paths:
        if filter_new:
            meta[key] = list(map( pathToTup, [P for P in generation_paths[key] if isNew(P)] ))
        elif filter_int:
            meta[key] = [TT for TT in map(pathToTup, generation_paths[key]) if TT[2][1] == 1]
        else:
            meta[key] = list( )
    return meta
#

    

In [16]:
'1101110'[-2:]

'10'

# Necklaces of 13

## Generation 7
'0110111' -> 13

0110111, 1101110, 1011101, 0111011, 1110110, 1101101, 1011011

```
('0110111', (7, 2, 11), (13, 1))
('0111011', (7, 2, 19), (109, 9))
('1011011', (7, 2, 22), (106, 9))
('1011101', (7, 2, 38), (10, 1))
('1101101', (7, 2, 44), (28, 3))
('1101110', (7, 2, 76), (52, 9))
('1110110', (7, 2, 88), (40, 9))
```
Note: $(10,1)$ and $(28,3)$ would be found in generation 5 since they have `01` suffix

## Generation 9


In [17]:
CollatzNumNecklacesMeta(13, 2)

{7: [], 9: []}

In [18]:
CollatzNumNecklacesMeta(13, 20, filter_new=True)

{7: [('0110111', (7, 2, 11), (13, 1)),
  ('0111011', (7, 2, 19), (109, 9)),
  ('1011011', (7, 2, 22), (106, 9)),
  ('1101110', (7, 2, 76), (52, 9)),
  ('1110110', (7, 2, 88), (40, 9))],
 9: [('010110111', (9, 3, 53), (17, 1)),
  ('011101011', (9, 3, 121), (391, 27)),
  ('101011011', (9, 3, 106), (406, 27)),
  ('101101110', (9, 3, 322), (190, 27)),
  ('110111010', (9, 3, 484), (28, 27)),
  ('111010110', (9, 3, 424), (88, 27))],
 11: [('01010110111', (11, 4, 239), (67, 3)),
  ('01110101011', (11, 4, 619), (1429, 81)),
  ('10101011011', (11, 4, 478), (1570, 81)),
  ('10101101110', (11, 4, 1342), (706, 81)),
  ('10110111010', (11, 4, 1990), (58, 81)),
  ('11011101010', (11, 4, 2476), (-428, 81)),
  ('11101010110', (11, 4, 1912), (136, 81))],
 13: [('0101010110111', (13, 5, 1037), (265, 9)),
  ('0111010101011', (13, 5, 2881), (5311, 243)),
  ('1010101011011', (13, 5, 2074), (6118, 243)),
  ('1010101101110', (13, 5, 5530), (2662, 243)),
  ('1010110111010', (13, 5, 8122), (70, 243)),
  ('1011

In [19]:
CollatzNumNecklacesMeta(13, 80, filter_int=True)

{7: [('0110111', (7, 2, 11), (13, 1)), ('1011101', (7, 2, 38), (10, 1))],
 9: [('010110111', (9, 3, 53), (17, 1)),
  ('011011101', (9, 3, 161), (13, 1)),
  ('101110101', (9, 3, 242), (10, 1))],
 11: [('01011011101', (11, 4, 671), (17, 1)),
  ('01101110101', (11, 4, 995), (13, 1)),
  ('10111010101', (11, 4, 1238), (10, 1))],
 13: [('0101101110101', (13, 5, 4061), (17, 1)),
  ('0110111010101', (13, 5, 5033), (13, 1)),
  ('1011101010101', (13, 5, 5762), (10, 1))],
 15: [('010110111010101', (15, 6, 20375), (17, 1)),
  ('011011101010101', (15, 6, 23291), (13, 1)),
  ('101110101010101', (15, 6, 25478), (10, 1))],
 17: [('01011011101010101', (17, 7, 93893), (17, 1)),
  ('01101110101010101', (17, 7, 102641), (13, 1)),
  ('10111010101010101', (17, 7, 109202), (10, 1))],
 19: [('0101101110101010101', (19, 8, 412751), (17, 1)),
  ('0110111010101010101', (19, 8, 438995), (13, 1)),
  ('1011101010101010101', (19, 8, 458678), (10, 1))],
 21: [('010110111010101010101', (21, 9, 1762541), (17, 1)),
  ('

# Necklaces get "used up"?

We can generate necklaces that contain 17 forever and never pick up another integer? YES

The long chain of 1's this sort of expansion approach involves drives the solution to smaller numbers ...

In [20]:
-8712410196584 / 10460353203

-832.8982805365794

In [21]:
for tup in generation_tuples2(2):
    p2, p3, c = tup[1]
    fract = Fraction(2**p2 - c, 3**p3)
    print((tup[0], (p2, p3, c), (fract.numerator, fract.denominator)))

('11', (2, 0, 0), (4, 1))
('10', (2, 1, 2), (2, 3))
('01', (2, 1, 1), (1, 1))
('00', (2, 2, 5), (-1, 9))


In [22]:
for tup in generation_tuples2(4):
    p2, p3, c = tup[1]
    fract = Fraction(2**p2 - c, 3**p3)
    print((tup[0], (p2, p3, c), (fract.numerator, fract.denominator)))

('1111', (4, 0, 0), (16, 1))
('1110', (4, 1, 8), (8, 3))
('1101', (4, 1, 4), (4, 1))
('1100', (4, 2, 20), (-4, 9))
('1011', (4, 1, 2), (14, 3))
('1010', (4, 2, 14), (2, 9))
('1001', (4, 2, 10), (2, 3))
('1000', (4, 3, 38), (-22, 27))
('0111', (4, 1, 1), (5, 1))
('0110', (4, 2, 11), (5, 9))
('0101', (4, 2, 7), (1, 1))
('0100', (4, 3, 29), (-13, 27))
('0011', (4, 2, 5), (11, 9))
('0010', (4, 3, 23), (-7, 27))
('0001', (4, 3, 19), (-1, 9))
('0000', (4, 4, 65), (-49, 81))


In [23]:
16*81, 4*81

(1296, 324)

In [24]:
8*27, 9*-4

(216, -36)

In [25]:
14*27, 2*27

(378, 54)

In [26]:
2*9, -22*3

(18, -66)

In [27]:
L = []
for tup in generation_tuples2(2):
    p2, p3, c = tup[1]
    fract = Fraction(2**p2 - c, 3**p3)
    gennum = (2**p2 - c)*(3**(p2-p3))
    L.append((gennum, tup[0], (fract.numerator, fract.denominator)))
sorted(L)

[(-1, '00', (-1, 9)), (6, '10', (2, 3)), (9, '01', (1, 1)), (36, '11', (4, 1))]

In [28]:
L = []
for tup in generation_tuples2(4):
    p2, p3, c = tup[1]
    fract = Fraction(2**p2 - c, 3**p3)
    gennum = (2**p2 - c)*(3**(p2-p3))
    L.append((gennum, tup[0], (fract.numerator, fract.denominator)))
sorted(L)

[(-66, '1000', (-22, 27)),
 (-49, '0000', (-49, 81)),
 (-39, '0100', (-13, 27)),
 (-36, '1100', (-4, 9)),
 (-21, '0010', (-7, 27)),
 (-9, '0001', (-1, 9)),
 (18, '1010', (2, 9)),
 (45, '0110', (5, 9)),
 (54, '1001', (2, 3)),
 (81, '0101', (1, 1)),
 (99, '0011', (11, 9)),
 (216, '1110', (8, 3)),
 (324, '1101', (4, 1)),
 (378, '1011', (14, 3)),
 (405, '0111', (5, 1)),
 (1296, '1111', (16, 1))]

In [29]:
L = []
for tup in generation_tuples2(6):
    p2, p3, c = tup[1]
    fract = Fraction(2**p2 - c, 3**p3)
    gennum = (2**p2 - c)*(3**(p2-p3))
    L.append((gennum, tup[0], (fract.numerator, fract.denominator)))
sorted(L)

[(-2376, '111000', (-88, 27)),
 (-1764, '110000', (-196, 81)),
 (-1404, '110100', (-52, 27)),
 (-1296, '111100', (-16, 9)),
 (-1278, '101000', (-142, 81)),
 (-1074, '100000', (-358, 243)),
 (-1035, '011000', (-115, 81)),
 (-954, '100100', (-106, 81)),
 (-918, '101100', (-34, 27)),
 (-831, '010000', (-277, 243)),
 (-756, '110010', (-28, 27)),
 (-738, '100010', (-82, 81)),
 (-711, '010100', (-79, 81)),
 (-675, '011100', (-25, 27)),
 (-669, '001000', (-223, 243)),
 (-601, '000000', (-601, 729)),
 (-594, '100001', (-22, 27)),
 (-561, '000100', (-187, 243)),
 (-549, '001100', (-61, 81)),
 (-495, '010010', (-55, 81)),
 (-489, '000010', (-163, 243)),
 (-441, '000001', (-49, 81)),
 (-351, '010001', (-13, 27)),
 (-333, '001010', (-37, 81)),
 (-324, '110001', (-4, 9)),
 (-270, '101010', (-10, 27)),
 (-225, '000110', (-25, 81)),
 (-189, '001001', (-7, 27)),
 (-81, '000101', (-1, 9)),
 (-27, '011010', (-1, 27)),
 (-9, '000011', (-1, 81)),
 (54, '100110', (2, 27)),
 (162, '101001', (2, 9)),
 (297, 

In [30]:
FractionFromPath("010011100110010110111")

(89, 1)

In [35]:
CollatzPath(27), CollatzPath(54)

NameError: name 'CollatzPath' is not defined